Explain why we need to hash the palantir-java-format jar as part of SpotlessTask's state.

kelvinou01 opened the pull request:
## Before this PR
<!-- What's wrong with the current state of the world and why change it now? -->

FLUP to https://github.com/palantir/palantir-java-format/pull/1313#discussion_r2137623833

~~In `PalantirJavaFormatStep`, we currently sign the palantir-java-format jar file, and serialize that signature as part of the `State`. I think [the `State` is *only* used to create the format function](https://github.com/diffplug/spotless/blob/52654ef8c4a6191d983b10a2370d53b1ca023f7d/lib/src/main/java/com/diffplug/spotless/FormatterStepImpl.java#L80), and nothing else like up-to-date checking. However, seems like the signature doesn't do anything, and just lives as a serializable field in `State`.~~

~~If I am correct, this is wasted computation, repeated once per subproject in your gradle build.~~

Seems like I was wrong. This hashing is both 1. efficient and 2. necessary. Let's add a comment explaning our findings. 

## After this PR
<!-- User-facing outcomes this PR delivers go below -->
Explain why we need to hash the palantir-java-format jar as part of SpotlessTask's state. #1327

==COMMIT_MSG==
Explain why we need to hash the palantir-java-format jar as part of SpotlessTask's state
==COMMIT_MSG==

## Possible downsides?
<!-- Please describe any way users could be negatively affected by this PR. -->

FinlayRJW committed:
make lazy

Commit:
Make NativePalantirJavaFormatStep lazy as well

kelvinou01 committed:
Format

kelvinou01 committed:
Add test for configuration cache

kelvinou01 committed:
Add test for configuration cache

kelvinou01 committed:
Change test

kelvinou01 committed:
Change test

svc-changelog committed:
Add generated changelog entries

svc-changelog committed:
Add generated changelog entries

kelvinou01 committed:
Remove configuration cache test

kelvinou01 committed:
Remove configuration cache test

kelvinou01 committed:
Merge github.com:palantir/palantir-java-format into finlayw/cc

FinlayRJW committed:
finlay fixes

FinlayRJW committed:
finlay fixes

FinlayRJW committed:
finlay fixes

kelvinou01 committed:
Merge branch 'finlayw/cc' of github.com:palantir/palantir-java-format into finlayw/cc

FinlayRJW committed:
finlay tidy

kelvinou01 committed:
Clean things up

kelvinou01 committed:
Merge branch 'finlayw/cc' of github.com:palantir/palantir-java-format into finlayw/cc

kelvinou01 committed:
Remove dumb comment

svc-changelog committed:
Add generated changelog entries

svc-changelog committed:
Add generated changelog entries

svc-changelog committed:
Add generated changelog entries

svc-changelog committed:
Add generated changelog entries

kelvinou01 committed:
Fix release note

kelvinou01 committed:
Fix release note

kelvinou01 committed:
Add test

kelvinou01 committed:
Add test

svc-changelog committed:
Add generated changelog entries

svc-changelog committed:
Add generated changelog entries

svc-changelog committed:
Add generated changelog entries

svc-changelog committed:
Add generated changelog entries

svc-changelog committed:
Add generated changelog entries

svc-changelog committed:
Add generated changelog entries

kelvinou01 committed:
Attempted debugging

kelvinou01 committed:
Attempted debugging

kelvinou01 committed:
Work state

kelvinou01 committed:
Work state

kelvinou01 committed:
Work state

kelvinou01 committed:
Work state

kelvinou01 committed:
Merge branch 'develop' of github.com:palantir/palantir-java-format into okelvin/cc-test

kelvinou01 committed:
its WORKING

kelvinou01 committed:
its WORKING

kelvinou01 committed:
Merge branch 'okelvin/cc-test' of github.com:palantir/palantir-java-format into finlayw/cc

# Conflicts:
#	gradle-palantir-java-format/src/main/java/com/palantir/javaformat/gradle/spotless/NativePalantirJavaFormatStep.java

kelvinou01 committed:
Make it pretty, make it nice

kelvinou01 committed:
Make it pretty, make it nice

kelvinou01 committed:
Fix versions

kelvinou01 committed:
Fix versions

kelvinou01 committed:
Do classes instead

kelvinou01 committed:
Do classes instead

kelvinou01 committed:
Merge branch 'develop' of github.com:palantir/palantir-java-format into finlayw/cc

# Conflicts:
#	gradle-palantir-java-format/src/main/java/com/palantir/javaformat/gradle/spotless/NativePalantirJavaFormatStep.java

kelvinou01 committed:
Remove imports

kelvinou01 committed:
Minimize test

kelvinou01 committed:
Minimize test

FinlayRJW committed:
check getByName regression

FinlayRJW committed:
Merge branch 'finlayw/cc' of github.com:palantir/palantir-java-format into finlayw/cc

kelvinou01 committed:
Plz work

kelvinou01 committed:
Remove old tests

kelvinou01 committed:
Remove old tests

FinlayRJW committed:
remove finlay messing bits

kelvinou01 committed:
Polish docs

kelvinou01 committed:
Polish docs

kelvinou01 committed:
Fix this

kelvinou01 committed:
Merge branch 'finlayw/cc' of github.com:palantir/palantir-java-format into finlayw/cc

kelvinou01 committed:
Add a detailed javadoc

kelvinou01 committed:
Make it pretty

kelvinou01 committed:
Make it pretty

kelvinou01 committed:
Remove unused method

kelvinou01 committed:
Remove unused method

kelvinou01 committed:
boolean

kelvinou01 committed:
boolean

kelvinou01 committed:
Update gradle-palantir-java-format/src/test/groovy/com/palantir/javaformat/gradle/GradlewExecutor.groovy

Co-authored-by: Felix de Souza <desouza.felix.93@gmail.com>

kelvinou01 committed:
Update gradle-palantir-java-format/src/test/groovy/com/palantir/javaformat/gradle/GradlewExecutor.groovy

Co-authored-by: Felix de Souza <desouza.felix.93@gmail.com>

kelvinou01 committed:
Update gradle-palantir-java-format/src/test/groovy/com/palantir/javaformat/gradle/SupportsSpotless622.groovy

Co-authored-by: Felix de Souza <desouza.felix.93@gmail.com>

kelvinou01 committed:
Merge github.com:palantir/palantir-java-format into finlayw/cc

kelvinou01 committed:
Merge branch 'finlayw/cc' of github.com:palantir/palantir-java-format into finlayw/cc

kelvinou01 committed:
Update docstring formatting

kelvinou01 committed:
Remove extra assert

kelvinou01 committed:
Merge branch 'develop' of github.com:palantir/palantir-java-format into finlayw/cc

# Conflicts:
#	gradle-palantir-java-format/src/main/java/com/palantir/javaformat/gradle/spotless/NativePalantirJavaFormatStep.java

svc-excavator-bot committed:
Excavator:  Upgrades Baseline to the latest version (#1326)

CRogers commented on gradle-palantir-java-format/src/main/java/com/palantir/javaformat/gradle/spotless/PalantirJavaFormatStep.java:
```diff
@@ -53,34 +49,19 @@ static final class State implements Serializable {
         @SuppressWarnings({"unused", "FieldCanBeLocal"})
         private FileSignature jarsSignature;
```
This doesn't seem to be used anywhere - can we just remove it?

CRogers replied:
Even before it was not used - I don't get it.

CRogers replied:
Oh, it was created in createFormat() - seems very odd.

CRogers replied:
Ok, so I looked a bit deeper, and spotless has created this crazy system using Java Serialisation (which Gradle sort of understands) and this eventually becomes part of the input to `@Input SpotlessTask#getSteps`, so this is a Gradle input. We probably want to keep something here that represents the hashes of all files, so if this changed the task would not be up-to-date.

This is called signature, but it's not actually making a cryptographic signature - it's just [doing the name/size/hash](https://github.com/diffplug/spotless/blob/228eb10af382b19e130d8d9479f7a95238cb4358/lib/src/main/java/com/diffplug/spotless/FileSignature.java#L181-L186). The hash is [sha-256](https://github.com/diffplug/spotless/blob/228eb10af382b19e130d8d9479f7a95238cb4358/lib/src/main/java/com/diffplug/spotless/FileSignature.java#L151).

Spotless also has [caching](https://github.com/CRogers/spotless/blob/228eb10af382b19e130d8d9479f7a95238cb4358/lib/src/main/java/com/diffplug/spotless/FileSignature.java#L138-L143) so this won't be recalculated multiple times.

All in all I think we should leave this. Maybe make so notes about what we've learned here as a comment at most

kelvinou01 replied:
Yep, I've added a comment explaining what we've found. 

> Gradle understands Java's `Serializable`, and uses it to invalidate `@Input`s to tasks.

Is this claim correct? I'm not sure where's the source code/docs. Is it [this](https://docs.gradle.org/current/userguide/configuration_cache.html#:~:text=Gradle%20uses%20an%20optimized%20serialization,simple%20state%20or%20supported%20types.)?

CRogers reviewed (dismissed):
Thanks for adding this explanation

CRogers commented on gradle-palantir-java-format/src/main/java/com/palantir/javaformat/gradle/spotless/PalantirJavaFormatStep.java:
```diff
@@ -30,7 +30,8 @@ public final class PalantirJavaFormatStep {
     private static final String IMPL_CLASS = "com.palantir.javaformat.java.Formatter";
 
-    private PalantirJavaFormatStep() {}
+    private PalantirJavaFormatStep() {
+    }
```
Just need to fix the formatting issues like this and the too long comments.

svc-excavator-bot committed:
Excavator:  Upgrades Baseline to the latest version (#1328)

kelvinou01 commented on gradle-palantir-java-format/src/main/java/com/palantir/javaformat/gradle/spotless/PalantirJavaFormatStep.java:
```diff
@@ -49,7 +50,16 @@ static final class State implements Serializable {
         private final String stepName = NAME;
 
-        // Kept for state serialization purposes.
+        // Spotless' `FormatterStepImpl` implements Java's `Serializable` interface in a weird way:
```
Add a similar explanation for NativePalantirJavaFormatStep

kelvinou01 committed:
Rename .groovy to .java

kelvinou01 committed:
Rename .groovy to .java

kelvinou01 committed:
Add comments

kelvinou01 committed:
Propagate

kelvinou01 committed:
Propagate

kelvinou01 committed:
Add comments

# Conflicts:
#	gradle-palantir-java-format/src/main/java/com/palantir/javaformat/gradle/spotless/PalantirJavaFormatStep.java

kelvinou01 committed:
Add comments

# Conflicts:
#	gradle-palantir-java-format/src/main/java/com/palantir/javaformat/gradle/spotless/PalantirJavaFormatStep.java

kelvinou01 committed:
Add comment for native stpe

kelvinou01 committed:
Add comment for native stpe

kelvinou01 committed:
Remove diffs

kelvinou01 committed:
Remove diffs

kelvinou01 committed:
Fix stuff

kelvinou01 committed:
Fix stuff

kelvinou01 committed:
Merge branch 'okelvin/no-sign' of github.com:palantir/palantir-java-format into okelvin/no-sign

# Conflicts:
#	gradle-palantir-java-format/src/main/java/com/palantir/javaformat/gradle/spotless/NativePalantirJavaFormatStep.java

kelvinou01 committed:
sTaNdArDs

kelvinou01 committed:
Delete changelog/@unreleased/pr-1313.v2.yml

kelvinou01 committed:
Weird

kelvinou01 committed:
Merge branch 'okelvin/no-sign' of github.com:palantir/palantir-java-format into okelvin/no-sign

felixdesouza commented on gradle-palantir-java-format/src/main/java/com/palantir/javaformat/gradle/spotless/PalantirJavaFormatStep.java:
```diff
@@ -75,7 +86,11 @@ FormatterFunc createFormat() {
                     Iterable<File> jars = jarsSupplier.get();
+
+                    // Not a performance issue, as Spotless caches this
+                    // https://github.com/CRogers/spotless/blob/228eb10af382b19e130d8d9479f7a95238cb4358/lib/src/main/java/com/diffplug/spotless/FileSignature.java#L138-L143
```
nit: can this point to `diffplug`?

kelvinou01 replied:
gotcha!

kelvinou01 committed:
fix diffplug
